// kea-typegen/dist/__tests__/e2e/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../utils");
test('actions - with a function', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: () => ({
                updateName: (name: string) => ({ name }),
                updateOtherName: (otherName: string) => ({ otherName }),
            })
        })
    `;
    expect(utils_1.logicSourceToLogicType(logicSource)).toMatchSnapshot();
});
test('actions - as an object', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: {
                updateName: (name: string) => ({ name }),
                updateOtherName: (otherName: string) => ({ otherName }),
            }
        })
    `;
    expect(utils_1.logicSourceToLogicType(logicSource)).toMatchSnapshot();
});
test('actions - as a function returning a object', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: function () {
                return {
                    updateName: (name: string) => ({ name }),
                    updateOtherName: (otherName: string) => ({ otherName }),
                }
            }
        })
    `;
    expect(utils_1.logicSourceToLogicType(logicSource)).toMatchSnapshot();
});
test('actions - with random values instead of functions', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: {
                updateName: (name?: string) => ({ name }),
                withDefaultValue: (name = "john") => ({ name }),
                withDefaultValueAndType: (name: string = "john") => ({ name }),
                withBool: true,
                withRandomPayload: { bla: 123 },
            }
        })
    `;
    expect(utils_1.logicSourceToLogicType(logicSource)).toMatchSnapshot();
});
